Edits from Semih (#78)
kuzudb/kuzu-blog: src/content/post/2024-08-15-kuzu-v-0.5.1.md

tags: ["cypher"]
---

This release comes with several bug fixes and a much awaited feature: in-memory mode for Kùzu to quickly create temporary and in-memory databases!

> Please see the [release notes](https://github.com/kuzudb/kuzu/releases/tag/v0.5.1) for the bugfixes. In the rest of this post,

Reviewer: @prrao87 can u also change the link here to `https://github.com/kuzudb/kuzu/releases/tag/v0.6.0`? I've switched the branch from 0.5.1 to 0.6.0 in main repo.

Reviewer 2: Yes, will do.

we give an overview of the in-memory Kùzu feature and provide some insights about the performance benefits to expect
when using Kùzu in in-memory mode. 

## In-memory Mode
Kùzu now supports both on-disk and in-memory modes. 
As you create your databases, if you do not specify a database path or specify it as `:memory:`, Kùzu will be opened under **in-memory** mode.

> The main differences of using in-memory on-disk mode compared to the on-disk mode are:

Reviewer 2: ```suggestion
The main differences of using in-memory mode compared to the on-disk mode are:
```

> - There are **no writes to the write-ahead-log (WAL) during transactions**, no data is persisted to disk (so `CHECKPOINT` will do nothing); 

Reviewer 2: ```suggestion
- There are **no writes to the write-ahead-log (WAL) during transactions**, so no data is persisted to disk (so `CHECKPOINT` will do nothing); 
```

```cypher
// COPY
COPY Comment FROM 'ldbc/ldbc-100/csv/comment_0_0.csv' (DELIM="|", HEADER=true);
// Insert
CREATE NODE TABLE nodeT(id INT64, name STRING, age INT64, net_worth FLOAT, PRIMARY KEY (id));
// Pass each record's values as parameters as an individual transaction
CREATE (:nodeT {id: $id, name: $name, age: $age, net_worth: $net_worth})
// Scan
MATCH (c:Comment) RETURN MIN(c.ID), MIN(c.creationDate), MIN(c.locationIP), MIN(c.browserUsed), MIN(c.content), MIN(c.length);
// 2-hop Join
MATCH (a:Person)-[:knows]->(b:Person)-[:knows]->(c:Person) RETURN MIN(a.birthday), MIN(b.birthday), MIN(c.birthday);
```

| mode       |        on-disk (ms)       |     in-mem (ms)   |
| ---------- | ------------------------- | ----------------- |
| COPY       |     34586                 | 14798             |
| Insert     | 79313                     | 47814             |
| Scan       | 5330 (cold) / 1801 (warm) | 1896              |
| 2-hop Join |  950 (cold) / 899 (warm)  | 904               |

The key takeaways are:
- The performance of COPY and large scans during cold runs are much improved (from 34.5s to 14.7s in this experiment) under the in-memory mode compared to the on-disk mode. 

> This is due to avoiding all disk I/Os that happen that the on-disk mode has to do to persist the data.

Reviewer 2: ```suggestion
This is due to avoiding all disk I/Os that the on-disk mode has to do to persist the data.
```

- Similarly, the performance of insertions is greatly improved (from 79.3s to 47.8s in this experiment) due to no writes to the WAL, which involves writing and syncing the disk file.

> - Scans: ``Cold'' scans, i.e., the initial scans that are done when the database starts and the buffer manager is empty also improve significantly (from 5.3s to 1.8s). For warm scans, the performance is empty.   

Reviewer 2: ```suggestion
- Scans: "Cold" scans, i.e., the initial scans that are done when the database starts and the buffer manager is empty, also improve significantly (from 5.3s to 1.8s). For warm scans, the performance is empty.   
```

Reviewer 2: @ray6080 I presume this last sentence should say "For warm scans, the performance is difference is negligible".

Reviewer: yeah. I think so too.

> - For large scans during warm runs, the performance difference between the two modes is neligible (*1896ms vs. 1801ms*), since needed pages are already cached in our buffer manager.

Reviewer 2: ```suggestion
- For large scans during warm runs, the performance difference between the two modes is neligible (*1896ms vs. 1801ms*), since the required pages are already cached in the buffer manager.
```

- For 2-hop joins, where the performance bottleneck is in the joins and not the scans, thus performance difference is small in both cold and warm runs of the query.

> So, you can expect the in-memory mode to improve the performance of your data ingestion pipelines, such as a COPY statement or

Reviewer 2: ```suggestion
Overall, you can expect the in-memory mode to improve the performance of your data ingestion pipelines, such as a COPY statement or
```

your write transaction-heavy workloads. You can also expect visible performance improvements if you are running a query only once before
closing your database. So a scenario where you create a temporary database and run a few queries is a good case where
good performance improvements can be expected.

## Closing Marks

> Many users had asked for this feature and we hope this simplifies and possibly speeds up some of your workloads.

Reviewer 2: ```suggestion
The in-memory feature is the first of many more that we have in our upcoming roadmap for Kùzu. Many users had asked for [this feature](https://github.com/kuzudb/kuzu/issues/1816), so we hope it simplifies and possibly speeds up some of your workloads.
```

> See the documentation [here](xxx) for more details of how to create an in-memory database. Finally, come to our Discord 

Reviewer 2: ```suggestion
See the documentation [here](xxx) for more details of how to create an in-memory database. Try out the new version and its remaining features, and come over to our Discord 
```
